File: lib/tracker/progress.ts
export interface StreakInput {
  current: number;
  best: number;
  yesterdayCleared: boolean;
}

export interface StreakResult {
  current: number;
  best: number;
}

/** Pure: streak transition for the moment today's required quests all clear. */
export function decideStreak(input: StreakInput): StreakResult {
  const current = input.yesterdayCleared ? input.current + 1 : 1;
  return { current, best: Math.max(current, input.best) };
}

File: api/app/logic/progress.py
"""Streak transition — port of lib/tracker/progress.ts."""

from __future__ import annotations

from dataclasses import dataclass


@dataclass
class StreakResult:
    current: int
    best: int


def decide_streak(*, current: int, best: int, yesterday_cleared: bool) -> StreakResult:
    """Streak transition for the moment today's required quests all clear."""
    new_current = current + 1 if yesterday_cleared else 1
    return StreakResult(current=new_current, best=max(new_current, best))
